articles/external_qos_file.md

```diff
@@ -0,0 +1,679 @@
+- Image pipeline nodes use some parameters to allow changing some of the policies ([PR](https://github.com/ros-perception/image_pipeline/pull/521)).
+- Ouster drivers also use some parameters to allow changing some policies ([PR](https://github.com/ros-drivers/ros2_ouster_drivers/pull/26)).
+- Gazebo ROS packages allows configuring QoS profiles of the plugins in the SDF file ([issue](https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1079)).
+
+## QoS file format
+
+DDS uses the concept of QoS profiles, in which each profile has a name.
+When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the QoS profiles file being used, and the QoS can be loaded in that way rather than from hard-coded values in the source code.
+
+If ROS would have QoS profiles names, like in DDS, collisions between those names in nodes written by different authors could happen.
+Thus, QoS profiles names would need the full power of remapping and expansion.
+
+This proposal uses the opposite approach, identifying an entity leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup> and through the use of topic/service names.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is assumed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+QoS settings apply to entities that are associated with ROS nodes.
+Thus, a top level tag to identify the node should be available in the file format.
+Here are a couple of examples:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        ...
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        ...
+```
+
+### Entity type
+
+It should be possible to distinguish between the different entity types in a QoS profile file.
+For example:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">  <!--Relative name-->
+            ...
+        </publisher>
+        <subscription topic_name="/another_ns/asd">  <!--Absolute name-->
+            ...
+        </subscription>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                ...
+        subscription:
+            topic_name: /another_ns/asd
+            qos:
+                ...
+```
+
+Besides `publisher` and `subscription` tags, `client` and `service` should be allowed.
+
+### QoS profiles IDs (optional)
+
+In the case that a node wants to create two `publishers` on the same topic with different QoS, the above format wouldn't allow to uniquely identify the `publisher`. The same applies to other kinds of entitys.
+
```

Comment:
> So the profile ID provides the matching between entities in the code and in the QoS configuration. I'm wondering if maybe there will be other configurations for entities in future and if it would make sense to have a more generic entity ID.

The concept of ID would apply to subscription/publishers/etc, I'm not sure if I understand the question.

Reply:
> 
> > So the profile ID provides the matching between entities in the code and in the QoS configuration. I'm wondering if maybe there will be other configurations for entities in future and if it would make sense to have a more generic entity ID.
> 
> The concept of ID would apply to subscription/publishers/etc, I'm not sure if I understand the question.

I more thought of other non-QoS topics. Maybe there will be things in future where you have to configure something per entity and a differentiation is needed. 
One example that just comes in my mind would be priorities for callbacks of subscribers. Imagine there is an executor that supports this and can also be configured similar to your QoS settings. Then you could also have two different subscribers on the same topic that shall be handled differently and must be uniquely identified (Maybe a crazy example). 
My point is, if you have to introduce an ID to solve this uniqueness issue for QoS, would it maybe make sense to introduce this more generally? Then it could be reused if other use cases come up where this is needed

Reply:
Ah, I see your point now.
I think it could be available for other things, there's nothing restricting that.
I wouldn't make that "id" introspectable in the graph though.